### ISSUE TITLE
Do losses (ex: dice_loss and focal_loss) work with multiple-class segmentation?

### ISSUE THREAD

**Issue by Minseok-Sakong** [mlyg/mixed-focal-loss#13]
I would like to use the losses for multi class segmentation using 3D Unet.
However, when performing binary segmentation, focal_loss seems to work well.
However, when I tried segmenting multi-class (3 classes), suddenly my model failed to learn.
The loss did not decrease and metrics (mean_IOU) did not change during the training.

Could you help me with the issue?
I have tried modifying the loss function after looking at your suggestion, here: https://github.com/mlyg/unified-focal-loss/issues/8.

My modified loss function for 3 labels:

      def categorical_focal_tversky_loss(delta=0.7, gamma=0.75):
          """This is the implementation for multiclass segmentation.
          Parameters
          ----------
          delta : float, optional
              controls weight given to false positive and false negatives, by default 0.7
          gamma : float, optional
              focal parameter controls degree of down-weighting of easy examples, by default 0.75
          """
          def loss_function(y_true, y_pred):
              # Clip values to prevent division by zero error
              epsilon = K.epsilon()
              y_pred = K.clip(y_pred, epsilon, 1. - epsilon)
      
              axis = identify_axis(y_true.get_shape())
              # Calculate true positives (tp), false negatives (fn) and false positives (fp)     
              tp = K.sum(y_true * y_pred, axis=axis)
              fn = K.sum(y_true * (1-y_pred), axis=axis)
              fp = K.sum((1-y_true) * y_pred, axis=axis)
              dice_class = (tp + epsilon)/(tp + delta*fn + (1-delta)*fp + epsilon)
      
              #calculate losses separately for each class, enhancing both classes
              back_dice = (1-dice_class[:,0]) * K.pow(1-dice_class[:,0], -gamma) 
              edge_dice = (1-dice_class[:,1]) * K.pow(1-dice_class[:,1], -gamma)
              brain_dice = (1-dice_class[:,2]) * K.pow(1-dice_class[:,2], -gamma)  
      
              # Average class scores
              loss = K.mean(tf.stack([back_dice,edge_dice, brain_dice],axis=-1))
              return loss
      
          return loss_function

**Comment by mlyg** [comment]
Hi Minseok-Sakong,

Thanks for your question. Your modification of the loss function looks correct to me, and is in fact identical (besides naming) to the code I used for evaluating on the KiTS19 dataset (3D multiclass).

Have you tried using a different loss function and seeing whether you can train the model? You can try the Dice loss in this repository (I think it should work without modification). That will help narrow down whether the issue is with the loss function or elsewhere.

**Comment by Minseok-Sakong** [comment]
Hi mlyg,
Thank you for the reply.
I investigated the problem and the problem was due to a change in the method loading up image datasets.
OpenCV did weird modifications to image pixel values.
I can confirm your losses work well with multi-class segmentation also.  Your loss functions are greatly valuable!
Thank you so much for your great work!

**Comment by mlyg** [comment]
I am very glad to hear that you fixed the issue and thank you very much for using this work!
